rfcs/convergence/context-values.md

```diff
@@ -0,0 +1,227 @@
+# RFC: Handling React context values in components
+
+---
+
+@layershifter @bsunderhus
+
+## Summary
+
+This RFC affects converged components and proposes multiple options detailing how to handle memoized context values for React Contexts.
+
+## Background
+
+There are cases when we need to pass some `props` or parts of `state` to child components, for this purpose we are using [React Context](https://reactjs.org/docs/context.html). To avoid unnecessary updates during components rerenders we are using either [context selector approach](https://github.com/dai-shi/use-context-selector) or [memoizing context value](https://reactjs.org/docs/context.html#contextprovider).
+
+```tsx
+// 💡 an example of memoization approach
+
+function Baz(props) {
+  const { inline, size } = props;
+  const value = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  // consumers of "SampleContext" will be notified only when "value" changes
+  return <SampleContext.Provider value={value} />;
+}
+```
+
+Currently, a few components (at least `Accordion`, `AccordionHeader`, `AccordionItem`) are defining a value for React Context on its state to use it in `render*` functions:
+
+```tsx
+// ⚠ simplified example
+
+function useBazState(props) {
+  const { inline, size } = props;
+  const [open, setOpen] = React.useState();
+
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+  const state = {
+    inline,
+    size,
+    open,
+    // 👇 "contextValue" is a part of component's state
+    contextValue,
+  };
+
+  return state;
+}
+
+function renderBaz(state) {
+  // 👇 "value" is memoized, consumer components will not rerender without need
+  return (
+    <SampleContext.Provider value={state.contextValue}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+```
+
+## Problem statement
+
+The problem comes from multiple places:
+
+- (_design issue_) `contextValue` is not a part of `state`, it's useless from component's perspective
+- (_usage issue_) in a case of composition state mutations are not intuitive
+
+The last item is a really serious problem as customers should be aware about our implementation details 🕵️‍
```

Comment:
customers should be aware about our implementation details? Can you elaborate ? Are we aware about implementation details of React and such ?

Reply:
Please check an example below, to override a single value if current state of things they will need to override it in two places.